src/dpl/src/Place.cpp

```diff
@@ -932,6 +933,112 @@ bool Opendp::checkRegionOverlap(const Cell* cell,
   // be fully contained by the cell's bounding box.
   return result.empty();
 }
+namespace cell_edges {
+Rect transformEdgeRect(const Rect& edge_rect,
+                       const Cell* cell,
+                       const DbuX x,
+                       const DbuY y,
+                       const odb::dbOrientType& orient)
+{
+  Rect bbox;
+  cell->db_inst_->getMaster()->getPlacementBoundary(bbox);
+  odb::dbTransform transform(orient);
+  transform.apply(bbox);
+  Point offset(x.v - bbox.xMin(), y.v - bbox.yMin());
+  transform.setOffset(offset);
+  Rect result(edge_rect);
+  transform.apply(result);
+  return result;
+}
+
+};  // namespace cell_edges
+bool Opendp::checkEdgeSpacing(const Cell* cell,
+                              const GridX x,
+                              const GridY y,
+                              const odb::dbOrientType& orient) const
+{
+  auto spacing_rules = db_->getTech()->getCellEdgeSpacingTable();
+  if (spacing_rules.empty()) {
+    return true;
+  }
+  const auto& master = db_master_map_.at(cell->db_inst_->getMaster());
+  auto x_real = gridToDbu(x, grid_->getSiteWidth());
+  auto y_real = grid_->gridYToDbu(y);
+  for (const auto& edge1 : master.edges_) {
+    for (auto spacing_rule : db_->getTech()->getCellEdgeSpacingTable()) {
+      if (spacing_rule->isOptional() || spacing_rule->isSoft()) {
+        continue;
+      }
+      std::string second_edge_type;
+      if (spacing_rule->getFirstEdgeType() == edge1.getEdgeType()) {
+        second_edge_type = spacing_rule->getSecondEdgeType();
+      } else if (spacing_rule->getSecondEdgeType() == edge1.getEdgeType()) {
+        second_edge_type = spacing_rule->getFirstEdgeType();
+      } else if (spacing_rule->getFirstEdgeType() == "DEFAULT") {
+        second_edge_type = spacing_rule->getSecondEdgeType();
+      } else if (spacing_rule->getSecondEdgeType() == "DEFAULT") {
+        second_edge_type = spacing_rule->getFirstEdgeType();
+      } else {
+        continue;
+      }
+      auto spc = spacing_rule->getSpacing();
+      auto edge1_box = cell_edges::transformEdgeRect(
+          edge1.getBBox(), cell, x_real, y_real, orient);
+      Rect query_rect(edge1_box);
+      bool is_vertical_edge = edge1_box.getDir() == 0;
+      if (is_vertical_edge) {
+        // vertical edge
+        query_rect = query_rect.bloat(spc, odb::Orientation2D::Horizontal);
+      } else {
+        // horizontal edge
+        query_rect = query_rect.bloat(spc, odb::Orientation2D::Vertical);
+      }
+      auto xMin = grid_->gridX(DbuX(query_rect.xMin()));
+      auto xMax = grid_->gridEndX(DbuX(query_rect.xMax()));
+      auto yMin = grid_->gridEndY(DbuY(query_rect.yMin())) - 1;
+      auto yMax = grid_->gridEndY(DbuY(query_rect.yMax()));
+      for (GridY y1 = yMin; y1 <= yMax; y1++) {
+        for (GridX x1 = xMin; x1 <= xMax; x1++) {
```

Comment:
Please add comment above the for loop that explains at a high level what the intent of the following lines is.